bitcoin/src/pow.rs

```diff
@@ -0,0 +1,1530 @@
+
+    /// Computes the popular "difficulty" measure for mining.
+    ///
+    /// Difficulty represents how difficult the current target makes it to find a block, relative to
+    /// how difficult it would be at the highest possible target (highest target == lowest difficulty).
+    ///
+    /// For example, a difficulty of 6,695,826 means that at a given hash rate, it will, on average,
+    /// take ~6.6 million times as long to find a valid block as it would at a difficulty of 1, or
+    /// alternatively, it will take, again on average, ~6.6 million times as many hashes to find a
+    /// valid block
+    ///
+    /// # Note
+    ///
+    /// Difficulty is calculated using the following algorithm `max / current` where [max] is
+    /// defined for the Bitcoin network and `current` is the current [target] for this block. As
+    /// such, a low target implies a high difficulty. Since [`Target`] is represented as a 256 bit
+    /// integer but `difficulty()` returns only 128 bits this means for targets below approximately
+    /// `0xffff_ffff_ffff_ffff_ffff_ffff` `difficulty()` will saturate at `u128::MAX`.
+    ///
+    /// [max]: Target::max
+    /// [target]: crate::blockdata::block::BlockHeader::target
+    pub fn difficulty(&self) -> u128 {
+        let d = Target::MAX.0 / self.0;
+        d.saturating_to_u128()
+    }
+}
+do_impl!(Target);
+
+impl Div for Target {
+    type Output = Self;
+    fn div(self, rhs: Self) -> Self { Self(self.0 / rhs.0) }
+}
+
+/// Encoding of 256-bit target as 32-bit float.
+///
+/// This is used to encode a target into the block header. Satoshi made this part of consensus code
+/// in the original version of Bitcoin, likely copying an idea from OpenSSL.
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub struct CompactTarget(u32);
+
+impl CompactTarget {
+    /// Creates a [`CompactTarget`] from a consensus encoded `u32`.
+    pub fn from_consensus(bits: u32) -> Self { Self(bits) }
+
+    /// Returns the consensus encoded `u32` representation of this [`CompactTarget`].
+    pub fn to_consensus(self) -> u32 { self.0 }
+}
+
+impl From<Target> for CompactTarget {
+    fn from(t: Target) -> Self { t.to_compact() }
+}
+
+impl From<CompactTarget> for Target {
+    fn from(c: CompactTarget) -> Self { Target::from_compact(c) }
+}
+
+impl Encodable for CompactTarget {
+    #[inline]
+    fn consensus_encode<W: Write + ?Sized>(&self, w: &mut W) -> Result<usize, io::Error> {
+        self.0.consensus_encode(w)
+    }
+}
+
+impl Decodable for CompactTarget {
+    #[inline]
+    fn consensus_decode<R: Read + ?Sized>(r: &mut R) -> Result<Self, encode::Error> {
+        u32::consensus_decode(r).map(CompactTarget)
+    }
+}
+
+/// Big-endian 256 bit integer type.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+struct U256(u128, u128);
```

Comment:
A non-doc comment saying field `0` is high, `1` is low could be useful for outsiders.

Reply:
Added `// (high, low): u.0 contains the high bits, u.1 contains the low bits.`